lib/temporal/worker/activity.rb

```diff
@@ -0,0 +1,76 @@
+require 'temporal/worker/sync_worker'
+require 'temporal/worker/activity_task_processor'
+
+module Temporal
+  class Worker
+    class Activity
+      def initialize(core_worker, converter, executor)
+        @running = false
+        @worker = SyncWorker.new(core_worker)
+        @converter = converter
+        @executor = executor
+      end
+
+      def run(reactor)
+        @running = true
+
+        while running?
+          activity_task = worker.poll_activity_task
+          reactor.async do
+            if activity_task.start
+              handle_start_activity(activity_task.task_token, activity_task.start)
+            elsif activity_task.cancel
+              handle_cancel_activity(activity_task.task_token, activity_task.cancel)
+            end
+          end
+        end
+      end
+
+      def shutdown
+        @running = false
+        executor.shutdown
+        # TODO: core worker shutdown implementation pending
+      end
+
+      private
+
+      attr_reader :worker, :converter, :executor
+
+      def running?
+        @running
+      end
+
+      def process_activity_task(task)
```

Comment:
Does not ever seem to be called unless I'm missing it (will ignore most of this file, seems to be in flux). Also, does this block a thread while processing the task (that may be ok, just struggling to follow the indirections from process task to task processor process)?

Reply:
This is called from line 67. Because it is called from within an `async` reactor it will only block a Fiber, but not a thread. We then call `executor.schedule` to dispatch all potentially blocking code on a thread pool (or any other provided executor), which includes converters, codecs, interceptors and the activity code itself.

I've added the indirection here to keep a handle in the future to a running activity (similar to Python's `_RunningActivity`). Hope that makes sense.

Reply:
Oh, I see, part of `handle_start_activity`, got it.

> Because it is called from within an async reactor it will only block a Fiber, but not a thread.

To confirm, `queue.pop` does not block the thread?

Reply:
It does, unless executed in a reactor context (within `Async` or `Async::Reactor#run`, which is exactly what is happening here). In this case it just yields the control back to the reactor until the queue has something on it

Reply:
I see, that library alters the Ruby stdlib. A bit concerning that devs try to share code in an os-threaded or green-threaded environment (may seem fine in most cases, but sometimes you need to know whether you're in an "async reactor"). But I guess that's how they dev'd it.

Reply:
Actually it doesn't alter the stdlib, it just adds few methods (`Async` and `Sync`) to the global scope for easy access. Unlike say [polyphony](https://github.com/digital-fabric/polyphony) that completely replaces lots of thread primitives with it's own implementation (which is also the reason for not using it). The async context is kept in a `Fiber.scheduler` and is scoped to the thread the reactor was executed on.

The main reason why it works with the stdlib primitives (like Thread and Queue) is that they have added hooks to the Fiber scheduler in Ruby 3+. Basically every time a stdlib is about to block (`sleep`, wait on a socket, `Queue#pop`, etc) it lets the scheduler decide how to handle it. The default implementation blocks, while `async` just treats this as a context switch and keeps the reactor running, resuming the interrupt when stdlib notifies it of an unblock.

Reply:
Ah, I see, they don't alter the stdlib, Ruby has scheduler hooks into some of the stdlib. Question - is it clear which calls in the the Ruby stdlib use the scheduler internally?

Reply:
[Not very clear](https://ruby-doc.org/core-3.0.0/Fiber/SchedulerInterface.html) and you're better off reading the Ruby source. But so far everything that I could think of had the hooks.

However our Client does not use these, because the implementation comes from Rust. We might add the hooks manually, though, depending on if Rutie has these bindings.

Reply:
:+1: I admit my main motivation for asking was how to differentiate between threaded and non-threaded activities (makes a big difference on scale), but it sounds like it's transparent.